keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,175 @@
+- Automatically provide logic for developers to guarantee immutability of
+  specific fields.
+
+### Non-Goals
+
+There are a few non-goals:
+- OpenAPI has a notion of "readOnly", but this is only meant to simplify schema,
+  and one is never supposed to send a readOnly value. This is different because
+  we want "writeOnce", not "readOnly".
+
+## Proposal
+
+This proposal attempts to create a concept of immutable and to visit some of
+what this can mean, but doesn't intend to propose all the possible
+semantics. The proposal should not close the door to further improvements.
+
+### Semantics
+
+We'll define immutable as "writeOnly", which means that the fields can only be
+set at creation time, and can never be updated after. Attempts to update an
+immutable field will result in an error (as opposed to being ignored), though we
+could potentially add that semantics later.
+
+It's also important to note that this concept is orthogonal to field ownership
+and list associativity. Ideally we would manage to keep the concepts entirely
+separate. e.g. An atomic list is a different concept from an immutable list.
+
+There are different semantics for a field to be immutable, especially if we
+consider non-leaf fields like lists and maps.
+
+#### Field selection
+
+- Recursive:
+  - Scalar fields are immutable (they can't change). Changing the
+    "type" will also typically be considered an error: `14 -> 14.0`. The field
+    should be purely round-tripped and will be compared as-is.
+  - Lists and maps are immutable too, which means that none of their field can
+    change, and new fields are not accepted.
+- Recursive with addition and/or deletion:
+  - Scalar fields are still immutable,
+  - New/delete fields in maps and lists are tolerated.
+  - The added/deleted type would probably be "Recursive" after the first level
+    (addition/delete is not recursive).
+  - For associative lists, keys are already "immutable" in a way (similar to how
+    names are "immutables" for objects). Changing the keys would mean deleting
+    and creating a new item in the list, which would be permitted by these
+    rules.
+- Mutable: A field set as mutable could cancel the recursive immutability.
+
+#### On mutation of selected fields
+
+- Error: mutation of an immutable field returns an error and the request is
+  rejected. No mutation is performed at all.
+- Ignored: mutation of the field (and/or sub-fields) are completely ignored, and
+  not persisted. They do not trigger an error.
+
+#### Proposal
+
+The behavior on mutation of immutable field is orthogonal to selection of
+immutable fields. If we accepted these two semantics at the same time, we would
+have to either have two different markers (one for selection and one for
+behavior) or have a single field with multiple options.
+
+For the moment, we understand that "Recursive" is the most important and simple
+selection mechanism, and "error" is the most natural way to handle mutations.
```

Comment:
Do I understand this correctly that for now the other options (recursive with add/del, ignore) are not proposed as part of this kep (yet) and therefore there is no Go marker for them in the examples below?

Reply:
I think I rephrased that,  thanks